### PR TITLE
Raftstore peer

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Last part of peer.rs.

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
@coocood PTAL

File: tikv/raftstore/peer.go [hunk 1]
```diff
@@ -992,3 +1028,930 @@ func (p *Peer) sendRaftMessage(msg eraftpb.Message, trans chan<- *rspb.RaftMessa
...
+			}
+			apply := &ApplyTask{
+				RegionId: p.regionId,
+				Term:     p.Term(),
+				Entries:  committedEntries,
+			}
+			ctx.applyRouter.ScheduleTask(p.regionId, apply)
+		}
+
+		p.ApplyReads(ctx, ready)
```
Comment (coocood) [inline tikv/raftstore/peer.go]: Should move out of the `else` block.

File: tikv/raftstore/peer.go [hunk 2]
```diff
@@ -992,3 +1028,930 @@ func (p *Peer) sendRaftMessage(msg eraftpb.Message, trans chan<- *rspb.RaftMessa
...
+	readyReadCount := p.RaftGroup.Raft.ReadyReadCount()
+
+	if pendingReadCount == lastPendingReadCount && readyReadCount == lastReadyReadCount {
+		// The message gets dropped silently, can't be handled anymore.
+		NotifyStaleReq(p.Term(), cb)
+		return false
+	}
+
+	cmds := make([]*ReqCbPair, 0, 1)
+	cmds = append(cmds, &ReqCbPair{req, cb})
```
Comment (coocood) [inline tikv/raftstore/peer.go]: This is simpler.
```
cmds := []*ReqCbPair{&ReqCbPair{req, cb}}
```

File: tikv/raftstore/peer.go [hunk 3]
```diff
@@ -992,3 +1028,930 @@ func (p *Peer) sendRaftMessage(msg eraftpb.Message, trans chan<- *rspb.RaftMessa
...
+			return nil, err
+		}
+		ctx.insert(ProposalContext_PrepareMerge)
+	}
+
+	return ctx, nil
+}
+
+func (p *Peer) ProposeNormal(pollCtx *PollContext, req *raft_cmdpb.RaftCmdRequest) (uint64, error) {
+	if p.PendingMergeState != nil && (req.AdminRequest == nil || req.AdminRequest.CmdType != raft_cmdpb.AdminCmdType_RollbackMerge) {
```
Comment (coocood) [inline tikv/raftstore/peer.go]: We can call `req.AdminRequest.GetCmdType()` directly without checking if `AdminRequest` is nil.

File: tikv/raftstore/peer.go [hunk 4]
```diff
@@ -992,3 +1028,930 @@ func (p *Peer) sendRaftMessage(msg eraftpb.Message, trans chan<- *rspb.RaftMessa
...
+		return 0, err
+	}
+
+	if uint64(len(data)) > pollCtx.Cfg.RaftEntryMaxSize {
+		log.Errorf("entry is too large, entry size %v", len(data))
+		return 0, &ErrRaftEntryTooLarge{RegionId: p.regionId, EntrySize: uint64(len(data))}
+	}
+
+	proposeIndex := p.nextProposalIndex()
+	// Todo: after the raft implement propose with context, chagne to:
```
Comment (coocood) [inline tikv/raftstore/peer.go]: s/Todo/TODO

File: tikv/raftstore/peer.go [hunk 5]
```diff
@@ -992,3 +1028,930 @@ func (p *Peer) sendRaftMessage(msg eraftpb.Message, trans chan<- *rspb.RaftMessa
...
+		return 0, &ErrNotLeader{RegionId: p.regionId}
+	}
+
+	return proposeIndex, nil
+}
+
+// Return true if the transfer leader request is accepted.
+func (p *Peer) ProposeTransferLeader(ctx *PollContext, req *raft_cmdpb.RaftCmdRequest, cb Callback) bool {
+	transferLeader := getTransferLeaderCmd(req)
+	if transferLeader == nil {
```
Comment (coocood) [inline tikv/raftstore/peer.go]: We can remove this check, it will panic when we call `transferLeaer.Peer`.

File: tikv/raftstore/peer.go [hunk 6]
```diff
@@ -992,3 +1028,930 @@ func (p *Peer) sendRaftMessage(msg eraftpb.Message, trans chan<- *rspb.RaftMessa
...
+		return 0, err
+	}
+
+	data, err := req.Marshal()
+	if err != nil {
+		return 0, err
+	}
+
+	changePeer := GetChangePeerCmd(req)
+	if changePeer == nil {
```
Comment (coocood) [inline tikv/raftstore/peer.go]: ditto

File: tikv/raftstore/peer.go (hunk 3 above)
Comment (zhangjinpeng1987) [reply tikv/raftstore/peer.go]: No, if `AdminRequest` is nil, here will panic, but TiKV will not panic.

**Comment by coocood** [comment]
LGTM

**Pull request merged**